php-checks/src/main/java/org/sonar/php/checks/phpunit/NotDiscoverableTestCheck.java

```diff
@@ -0,0 +1,153 @@
+    tree.accept(callsFindVisitor);
+    internalCalledMethods = callsFindVisitor.calledFunctions;
+
+    super.visitPhpUnitTestCase(tree);
+  }
+
+  @Override
+  public void visitMethodDeclaration(MethodDeclarationTree tree) {
+    if (!isPhpUnitTestCase()) {
+      return;
+    }
+
+    if (!CheckUtils.isPublic(tree) && isMarkedAsTestMethod(tree)) {
+      context().newIssue(this, tree.name(), MESSAGE_VISIBLE);
+    } else if (CheckUtils.isPublic(tree)
+      && !isMarkedAsTestMethod(tree)
+      && !internalCalledMethods.contains(tree.name().text().toLowerCase(Locale.ROOT))
+      && !OVERRIDABLE_METHODS.contains(tree.name().text().toLowerCase(Locale.ROOT))
+      && methodContainsAssertions(tree)) {
+      context().newIssue(this, tree.name(), MESSAGE_MARKED);
+    }
+  }
+
+  private static boolean methodContainsAssertions(MethodDeclarationTree tree) {
+    AssertionsFindVisitor assertionsFindVisitor = new AssertionsFindVisitor();
+    tree.accept(assertionsFindVisitor);
+    return assertionsFindVisitor.hasFoundAssertion;
+  }
+
+  private static boolean isMarkedAsTestMethod(MethodDeclarationTree tree) {
+    return tree.name().text().startsWith("test") || CheckUtils.hasAnnotation(tree, "test");
+  }
+
+  private static class InternalCallsFindVisitor extends PhpUnitCheck {
+    private final Set<String> calledFunctions = new HashSet<>();
+
+    @Override
+    public void visitFunctionCall(FunctionCallTree tree) {
+      if (!isPhpUnitTestMethod()) {
+        return;
+      }
+
+      String functionName = getFunctionName(tree);
+      if (functionName != null && isInternalMethodCall(tree)) {
+        calledFunctions.add(functionName);
+      }
+
+      super.visitFunctionCall(tree);
+    }
+
+    private static String getFunctionName(FunctionCallTree tree) {
+      String functionName = CheckUtils.getLowerCaseFunctionName(tree);
+
+      if (functionName != null && functionName.contains("::")) {
+        functionName = functionName.substring(functionName.lastIndexOf("::") + 2);
+      }
+
+      return functionName;
+    }
+
+    private static boolean isInternalMethodCall(FunctionCallTree tree) {
+      if (!tree.callee().is(Tree.Kind.OBJECT_MEMBER_ACCESS, Tree.Kind.CLASS_MEMBER_ACCESS)) {
+        return false;
+      }
+
+      String objectString = ((MemberAccessTree) tree.callee()).object().toString().toLowerCase(Locale.ROOT);
+      return SELF_OBJECTS.contains(objectString);
+    }
+  }
+
+  private static class AssertionsFindVisitor extends PHPVisitorCheck {
+    private boolean hasFoundAssertion = false;
+
+    @Override
+    public void visitFunctionExpression(FunctionExpressionTree tree) {
+      // Do not visit nested anonymous functions
+    }
+
+    @Override
+    public void visitFunctionCall(FunctionCallTree tree) {
+      if (isAssertion(tree)
+        && TreeUtils.findAncestorWithKind(tree, ImmutableList.of(Tree.Kind.NEW_EXPRESSION)) == null) {
```

Comment:
You can make a static method import to reduce the length

Reply:
I think when there is only one or a few usages static imports hurt readability more than they help.